File: scripts/ingest/sources.py
# -*- coding: utf-8 -*-
"""The source manifest, Python half (SAK-434).

This is scripts/ingest/sources.mjs in Python: the same file, the same format,
the same rules. Read that module's header for why the manifest exists and why
`archives` and `builtFrom` are written by different events. In short:

    archives   what each upstream archive is pinned to now: URL, the version the
               archive itself declares, and the SHA-256 of the file as
               downloaded.
    builtFrom  one entry per ingest pass, keyed by the script: the archives that
               pass read and the files under src/data/generated it wrote,
               stamped when it wrote them.

An ingest calls verify_source() before it reads an archive and stops when the
bytes differ, unless --accept-source is passed. After it writes its outputs it
calls record_build(). src/data/source-manifest.test.ts fails when the two
sections disagree, which is what stops a committed reduction from claiming an
archive it was not cut from.

Both halves write the file with json.dumps(..., indent=2, sort_keys=True) and a
trailing newline, so it does not churn depending on which language last touched
it.
"""

import bz2
import datetime
import gzip
import hashlib
import io
import json
import os
import urllib.request
import zipfile

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.normpath(os.path.join(HERE, "..", ".."))
MANIFEST = os.path.join(REPO, "src", "data", "generated", "sources.json")


def _today():
    return datetime.date.today().isoformat()


def read_manifest():
    with open(MANIFEST, encoding="utf-8") as fh:
        return json.load(fh)


def write_manifest(manifest):
    text = json.dumps(manifest, ensure_ascii=False, indent=2, sort_keys=True) + "\n"
    with open(MANIFEST, "w", encoding="utf-8") as fh:
        fh.write(text)


def sha256_file(path):
    digest = hashlib.sha256()
    with open(path, "rb") as fh:
        for chunk in iter(lambda: fh.read(1 << 20), b""):
            digest.update(chunk)
    return digest.hexdigest()


def archive_path(archive_id, manifest=None):
    """Absolute path to where an archive lives in the ignored raw directory."""
    manifest = manifest or read_manifest()
    entry = manifest["archives"].get(archive_id)
    if entry is None:
        raise SystemExit(f'no archive "{archive_id}" in src/data/generated/sources.json')
    return os.path.join(REPO, entry["path"])


def ensure_archive(archive_id):
    """Put an archive on disk at its manifest path if it is not there, and
    return that path.

    The ingest reads from a file rather than straight from the network: a
    response body that is never written down cannot be hashed the same way
    twice, and the hash is the whole point. The raw directory is ignored by git,
    so nothing this downloads is committed.
    """
    manifest = read_manifest()
    entry = manifest["archives"].get(archive_id)
    if entry is None:
        raise SystemExit(f'no archive "{archive_id}" in src/data/generated/sources.json')
    file = os.path.join(REPO, entry["path"])
    if os.path.exists(file):
        return file
    if entry.get("committed"):
        raise SystemExit(f"{entry['name']} is committed at {entry['path']} but is missing")
    os.makedirs(os.path.dirname(file), exist_ok=True)
    print(f"downloading {entry['name']} from {entry['url']}")
    urllib.request.urlretrieve(entry["url"], file)
    return file


def open_archive(archive_id, zip_member=None):
    """A binary stream over an archive's content, decompressed as needed.

    The hash is always of the file as downloaded, so .gz, .bz2 and .zip are
    unpacked here rather than leaving a second, unpinned copy on disk.
    """
    path = archive_path(archive_id)
    if zip_member is not None:
        with zipfile.ZipFile(path) as zf:
            return io.BytesIO(zf.read(zip_member))
    if path.endswith(".gz"):
        return gzip.open(path, "rb")
    if path.endswith(".bz2"):
        return bz2.open(path, "rb")
    return open(path, "rb")


def open_archive_text(archive_id, encoding="utf-8", zip_member=None):
    """A line-iterable text stream over an archive's content."""
    return io.TextIOWrapper(
        open_archive(archive_id, zip_member=zip_member), encoding=encoding
    )


def read_archive_text(archive_id, encoding="utf-8", zip_member=None):
    """An archive's content as one string. KRADFILE is distributed in EUC-JP,
    which is why the encoding is a parameter rather than always UTF-8.
    """
    with open_archive(archive_id, zip_member=zip_member) as fh:
        return fh.read().decode(encoding)


def add_source_args(parser):
    """The one flag every ingest takes: accept a new archive and record it."""
    parser.add_argument(
        "--accept-source",
        action="store_true",
        help="record the archive on disk as the new pin and build from it",
    )


def verify_source(archive_id, path=None, accept=False, version=None):
    """Check the bytes about to be read against the manifest; return the digest.

    Without `accept` a difference stops the run: rebuilding the committed
    reduction from an archive nobody chose is the exact failure the pins cannot
    see. With `accept` the entry is rewritten to this file and the run goes on.
    """
    manifest = read_manifest()
    entry = manifest["archives"].get(archive_id)
    if entry is None:
        raise SystemExit(
            f'No archive "{archive_id}" is recorded in src/data/generated/sources.json. '
            "Add it there before an ingest can read it."
        )
    file = os.path.abspath(path) if path else os.path.join(REPO, entry["path"])
    if not os.path.exists(file):
        raise SystemExit(
            f"{entry['name']} is not on disk at {os.path.relpath(file, REPO)}.\n"
            f"Download it from {entry['url']} to that path. The raw directory is "
            "ignored by git on purpose; the archives are not committed."
        )
    digest = sha256_file(file)
    if digest == entry["sha256"]:
        return digest
    if not accept:
        raise SystemExit(
            f"{entry['name']} is not the archive src/data/generated was built from.\n"
            f"  recorded  {entry['sha256']}\n"
            f"            version {entry['version']}, recorded {entry['recordedAt']}\n"
            f"  on disk   {digest}\n"
            f"            {os.path.relpath(file, REPO)}\n"
            "Stopping rather than rebuilding the reduction from an archive nobody\n"
            "chose. If this newer archive is the one you want, rerun with\n"
            "--accept-source: that records it here and builds from it. Every other\n"
            f"pass that reads {archive_id} then has to be re-run too, and\n"
            "src/data/source-manifest.test.ts fails until it is."
        )
    entry = dict(entry)
    entry["sha256"] = digest
    entry["bytes"] = os.path.getsize(file)
    entry["version"] = version if version is not None else entry["version"]
    entry["recordedAt"] = _today()
    manifest["archives"][archive_id] = entry
    write_manifest(manifest)
    print(f"recorded {entry['name']}: {digest} ({entry['version']})")
    return digest


def record_build(script, outputs, archive_ids):
    """Record one ingest pass: the archives it read, the files it wrote."""
    manifest = read_manifest()
    archives = {}
    for archive_id in archive_ids:
        entry = manifest["archives"].get(archive_id)
        if entry is None:
            raise SystemExit(f'no archive "{archive_id}" to record')
        archives[archive_id] = entry["sha256"]
    built = manifest.setdefault("builtFrom", {})
    built[script] = {
        "outputs": sorted(outputs),
        "archives": archives,
        "recordedAt": _today(),
    }
    write_manifest(manifest)

File: scripts/ingest/sources.mjs
// The source manifest: what upstream archive each ingest read, and the exact
// bytes it read (SAK-434).
//
// WHY THIS EXISTS
// ===============
// src/data/source-pins.test.ts pins every taught fact to the committed
// reduction under src/data/generated. It cannot reach further than that,
// because the upstream archives are not in the repo, so a re-run of an ingest
// against a different JMdict, a different KANJIDIC2 or a newer Tatoeba export
// would rewrite the reduction and every pin would still pass. The pins stop at
// the file; the file's own provenance was unrecorded.
//
// This module is the missing half. src/data/generated/sources.json records, for
// each upstream archive, the URL it came from, the version or date the archive
// itself declares, and the SHA-256 of the file as downloaded. Every ingest
// checks the archive it is about to read against that record and stops when it
// differs. An intentional upgrade is one flag; a silent change is a stop.
//
// THE TWO SECTIONS, AND WHY THEY ARE SEPARATE
// ===========================================
// `archives` is what the upstreams are pinned to NOW. `builtFrom` is one entry
// per INGEST PASS, keyed by the script: which archives that pass read and which
// files under src/data/generated it wrote, stamped at the moment it wrote them.
//
// A pass, not a file, because several passes write one file. build.py cuts
// kanji.json from KANJIDIC2 + KRADFILE + JMdict and kanji-raw-readings.py then
// back-fills on/kun onto it from KANJIDIC2 alone. Keyed by file, the second
// pass would overwrite the first pass's record and the file would claim an
// input history it does not have; keyed by pass, each records only what it read
// and what it touched.
//
// Nothing writes both sections at once: --accept-source rewrites an `archives`
// entry, and only a real write of outputs rewrites that pass's `builtFrom`
// entry. So accepting a newer KANJIDIC2 while running one script leaves every
// OTHER pass that read KANJIDIC2 carrying the old hash, and
// src/data/source-manifest.test.ts fails until those passes are re-run. That is
// the drift this is for: a reduction cannot go on claiming an archive it was
// not built from.
//
// The Python half of the ingest uses scripts/ingest/sources.py, which reads and
// writes the same file in the same format. Keep the two in step.

import { createHash } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { gunzipSync, inflateRawSync } from "node:zlib";

const HERE = dirname(fileURLToPath(import.meta.url));
export const REPO = resolve(HERE, "..", "..");
export const MANIFEST = join(REPO, "src", "data", "generated", "sources.json");

/** Today, as the date-only string the manifest stores. */
function today() {
  return new Date().toISOString().slice(0, 10);
}

/** Key order is not meaningful in JSON, so the file is written with every
 * object's keys sorted. Python's json.dumps(sort_keys=True) does the same, which
 * is what lets both halves of the ingest write this file without churning it. */
function sortDeep(value) {
  if (Array.isArray(value)) return value.map(sortDeep);
  if (value && typeof value === "object") {
    const out = {};
    for (const key of Object.keys(value).sort()) out[key] = sortDeep(value[key]);
    return out;
  }
  return value;
}

export function readManifest() {
  return JSON.parse(readFileSync(MANIFEST, "utf8"));
}

export function writeManifest(manifest) {
  writeFileSync(MANIFEST, JSON.stringify(sortDeep(manifest), null, 2) + "\n");
}

export function sha256File(path) {
  return createHash("sha256").update(readFileSync(path)).digest("hex");
}

/** Where an archive lives when it was downloaded to the ignored raw directory.
 * Absolute, so a script can hand it straight to its reader. */
export function archivePath(id, manifest = readManifest()) {
  const entry = manifest.archives[id];
  if (!entry) throw new Error(`no archive "${id}" in ${relative(REPO, MANIFEST)}`);
  return join(REPO, entry.path);
}

/**
 * Put an archive on disk at its manifest path if it is not there already, and
 * return that path.
 *
 * The ingest reads FROM A FILE rather than straight from the network, because a
 * response body that is never written down cannot be hashed the same way twice,
 * and the hash is the whole point. The raw directory is ignored by git, so
 * nothing here is committed.
 */
export async function ensureArchive(id) {
  const manifest = readManifest();
  const entry = manifest.archives[id];
  if (!entry) throw new Error(`no archive "${id}" in ${relative(REPO, MANIFEST)}`);
  const file = join(REPO, entry.path);
  if (existsSync(file)) return file;
  if (entry.committed) {
    throw new Error(`${entry.name} is committed at ${entry.path} but is missing`);
  }
  mkdirSync(dirname(file), { recursive: true });
  console.log(`downloading ${entry.name} from ${entry.url}`);
  const res = await fetch(entry.url);
  if (!res.ok) throw new Error(`fetch failed: ${res.status} ${res.statusText}`);
  writeFileSync(file, Buffer.from(await res.arrayBuffer()));
  return file;
}

/** An archive's text, gunzipped when the pinned file is a .gz. The hash is
 * always of the file as downloaded, so the decompression happens here rather
 * than leaving a second, unpinned copy on disk. */
export function readArchiveText(id, manifest = readManifest()) {
  const file = archivePath(id, manifest);
  const bytes = readFileSync(file);
  return (file.endsWith(".gz") ? gunzipSync(bytes) : bytes).toString("utf8");
}

/**
 * Read a pinned .zip archive without unpacking it to disk.
 *
 * Dependency-free on purpose: the ingest scripts run under plain `node` with
 * nothing installed, and scripts/ingest/cejc_reading_frequency.py already reads
 * its nested XLSX the same way on the Python side. Only the one shape these
 * archives use is handled: no zip64 (the largest here is 22MB over 11,661
 * entries, both far under the 4GB / 65,535 limits) and store or deflate.
 *
 * Returns { has(name), read(name) }, where read gives a Buffer.
 */
export function openArchiveZip(id, manifest = readManifest()) {
  const buf = readFileSync(archivePath(id, manifest));
  let eocd = -1;
  for (let i = buf.length - 22; i >= 0 && i > buf.length - 22 - 0xffff; i--) {
    if (buf.readUInt32LE(i) === 0x06054b50) {
      eocd = i;
      break;
    }
  }
  if (eocd < 0) throw new Error(`${id}: no zip end-of-central-directory record`);
  const count = buf.readUInt16LE(eocd + 10);
  let at = buf.readUInt32LE(eocd + 16);
  const entries = new Map();
  for (let n = 0; n < count; n++) {
    if (buf.readUInt32LE(at) !== 0x02014b50) throw new Error(`${id}: bad zip directory entry`);
    const method = buf.readUInt16LE(at + 10);
    const compressedSize = buf.readUInt32LE(at + 20);
    const nameLen = buf.readUInt16LE(at + 28);
    const extraLen = buf.readUInt16LE(at + 30);
    const commentLen = buf.readUInt16LE(at + 32);
    const localAt = buf.readUInt32LE(at + 42);
    const name = buf.toString("utf8", at + 46, at + 46 + nameLen);
    entries.set(name, { method, compressedSize, localAt });
    at += 46 + nameLen + extraLen + commentLen;
  }
  const read = (name) => {
    const entry = entries.get(name);
    if (!entry) throw new Error(`${id}: no member "${name}"`);
    const start =
      entry.localAt + 30 + buf.readUInt16LE(entry.localAt + 26) + buf.readUInt16LE(entry.localAt + 28);
    const raw = buf.subarray(start, start + entry.compressedSize);
    if (entry.method === 0) return Buffer.from(raw);
    if (entry.method === 8) return inflateRawSync(raw);
    throw new Error(`${id}: member "${name}" uses compression method ${entry.method}`);
  };
  return { has: (name) => entries.has(name), read };
}

/** True when the argv carries the flag that lets an ingest record a new
 * archive. One flag for an intentional upgrade; without it a changed archive
 * stops the run. */
export function acceptSource(argv = process.argv) {
  return argv.includes("--accept-source");
}

/**
 * Check the bytes an ingest is about to read against the manifest.
 *
 * Returns the archive's SHA-256 on success, so the caller can hand it to
 * recordBuild. With `accept`, the manifest entry is rewritten to this file
 * instead and the new digest is returned.
 *
 * `version` lets a caller pass the version string it read out of the archive
 * itself (KANJIDIC2's database_version, JMdict's creation date). It is only
 * used when accepting; a plain check never rewrites anything.
 */
export function verifySource(id, { path, accept = false, version = null } = {}) {
  const manifest = readManifest();
  const entry = manifest.archives[id];
  if (!entry) {
    throw new Error(
      `No archive "${id}" is recorded in ${relative(REPO, MANIFEST)}. ` +
        `Add it there before an ingest can read it.`,
    );
  }
  const file = path ? resolve(path) : join(REPO, entry.path);
  if (!existsSync(file)) {
    throw new Error(
      `${entry.name} is not on disk at ${relative(REPO, file)}.\n` +
        `Download it from ${entry.url} to that path. The raw directory is ` +
        `ignored by git on purpose; the archives are not committed.`,
    );
  }
  const digest = sha256File(file);
  if (digest === entry.sha256) return digest;
  if (!accept) {
    throw new Error(
      `${entry.name} is not the archive src/data/generated was built from.\n` +
        `  recorded  ${entry.sha256}\n` +
        `            version ${entry.version}, recorded ${entry.recordedAt}\n` +
        `  on disk   ${digest}\n` +
        `            ${relative(REPO, file)}\n` +
        `Stopping rather than rebuilding the reduction from an archive nobody\n` +
        `chose. If this newer archive is the one you want, rerun with\n` +
        `--accept-source: that records it here and builds from it. Every other\n` +
        `pass that reads ${id} then has to be re-run too, and\n` +
        `src/data/source-manifest.test.ts fails until it is.`,
    );
  }
  manifest.archives[id] = {
    ...entry,
    sha256: digest,
    bytes: statSync(file).size,
    version: version ?? entry.version,
    recordedAt: today(),
  };
  writeManifest(manifest);
  console.log(`recorded ${entry.name}: ${digest} (${version ?? entry.version})`);
  return digest;
}

/**
 * Record one ingest pass: which archives it read, which files it wrote.
 *
 * `script` is the pass's path from the repo root and is the entry's key.
 * `outputs` are paths under src/data/generated. `ids` are archive ids. The
 * hashes stored are the manifest's current ones, so this is only ever called
 * after verifySource has agreed the bytes on disk are those.
 */
export function recordBuild(script, outputs, ids) {
  const manifest = readManifest();
  const archives = {};
  for (const id of ids) {
    const entry = manifest.archives[id];
    if (!entry) throw new Error(`no archive "${id}" to record`);
    archives[id] = entry.sha256;
  }
  manifest.builtFrom ??= {};
  manifest.builtFrom[script] = {
    outputs: [...outputs].sort(),
    archives,
    recordedAt: today(),
  };
  writeManifest(manifest);
}
